### PR TITLE
A little bit of Machine Learning: Playing with Google's Prediction API

### PR DESCRIPTION
**Pull request by matryer**
Initial draft of "A little bit of Machine Learning: Playing with Google's Prediction API"

### CODE REVIEW

File: content/advent-2016/machine-learning-google-prediction-api.md [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+* "Awful performance"
+* "Great job"
+* "Absolutely love this great API"
+
+Our human brains can do this easily, but with a sensible and appropriate set of
+training data, it is possible to teach a machine to do it too with impressive results.
+
+## Training the Prediction API
+
+Creating training data isn't easy, and is often the most delecate part of the process.
```
Comment (ernesto-jimenez) [inline content/advent-2016/machine-learning-google-prediction-api.md]: s/delecate/delicate/?

File: content/advent-2016/machine-learning-google-prediction-api.md [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+Yes, it is a little weird that we're embedding CSV into JSON, and I expect this will evolve in future versions
+of the API. But it is still pretty clear what's going on; we're asking the Google Prediction API to predict which
+language the `"Please predict which language this is"` sentence is in.
+
+## Building a little app in Go
+
+We will write a little command line tool that lets us query the model from the terminal. 
+
+As the example project for this article was being put together, it became clear that the lion share of the work
```
Comment (ernesto-jimenez) [inline content/advent-2016/machine-learning-google-prediction-api.md]: s/lion/lion's/

although most non-native english speakers are probably unfamiliar with "lion's share"

File: content/advent-2016/machine-learning-google-prediction-api.md [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+was actually in authorising the requests via OAuth. And the experience for a terminal app was less than desirable; you 
+had to copy and paste a URL into a browser, log in (or create an account), click Accept to allow access to the API, 
+then copy the access code from the query parameters of a failed redirect and paste it into the terminal.
+
+OAuth is a web based authorisation protocol through and through, and short of implementing an HTTP endpoint to
+receive the request, there isn't much we can do to make the experience any better. So instead, we are going to be
+a little cheeky and piggy back on [Google's Try it app hosted on App Engine](http://try-prediction.appspot.com).
```
Comment (ernesto-jimenez) [inline content/advent-2016/machine-learning-google-prediction-api.md]: Why not start an HTTP server locally, show the URL to the server which triggers OAuth2's implicit flow to get the access token, saves the final access token and stops the server?

File: content/advent-2016/machine-learning-google-prediction-api.md [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+func do(query string) (*response, error) {
+	values := url.Values{
+		"model":  []string{"Language Detection"},
+		"Phrase": []string{query},
+	}
+	res, err := http.PostForm("http://try-prediction.appspot.com/predict", values)
+	if err != nil {
+		return nil, err
+	}
+	var result response
```
Comment (ernesto-jimenez) [inline content/advent-2016/machine-learning-google-prediction-api.md]: remember to close the body :)

File: content/advent-2016/machine-learning-google-prediction-api.md [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+Try some more:
+
+```bash
+> Joyeuses fêtes à vous mes chers Gophers
+  ^^ That looks like 🇫🇷 French to me
+> Felices fiestas para todos los Gophers
+  ^^ That looks like 🇪🇸 Spanish to me
+```
+
+Thank you to my multi-lingual European friends for their generousity in helping me translate this phrase
```
Comment (bradleyfalzon) [inline content/advent-2016/machine-learning-google-prediction-api.md]: `generousity` should be `generosity`.

File: content/advent-2016/machine-learning-google-prediction-api.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+in spite of Brexit.
+
+## Conslusion
```
Comment (bradleyfalzon) [inline content/advent-2016/machine-learning-google-prediction-api.md]: `Conslusion` should be `Conclusion`.

File: content/advent-2016/machine-learning-google-prediction-api.md [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+	OutputLabel string
+	OutputMulti []struct {
+		Label string
+		Score string
+	}
+	OutputValue float64
+}
+```
+
+The API gives us more information than we need, including a confidence level for each possible category. So for
```
Comment (bradleyfalzon) [inline content/advent-2016/machine-learning-google-prediction-api.md]: Perhaps you should tie in the previous mentions of regression vs categorisation here? Eg `including a confidence level as this is a regression not categorisation`?

File: content/advent-2016/machine-learning-google-prediction-api.md (hunk 3 above)
Comment (matryer) [inline content/advent-2016/machine-learning-google-prediction-api.md]: Ends up being the bulk of the work.

File: content/advent-2016/machine-learning-google-prediction-api.md [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+```
+"This is great", "Happy"
+"This is awful", "Sad"
+"I love this",   "Happy"
+"I hate this",   "Sad"
+(plus lots more questions/answers)
+```
+
+We could then to ask the Google Prediction API to tell us whether each of the following phrases we either happy, or sad:
```
Comment (matryer) [inline content/advent-2016/machine-learning-google-prediction-api.md]: Yes

Comment (matryer) [inline content/advent-2016/machine-learning-google-prediction-api.md]: ok

**Pull request merged**